src/modules/esl-popup/core/esl-popup.ts

```diff
@@ -1,26 +1,168 @@
 import {ExportNs} from '../../esl-utils/environment/export-ns';
-import {attr} from '../../esl-base-element/core';
+import {attr, jsonAttr} from '../../esl-base-element/core';
+import {prop} from '../../esl-utils/decorators/prop';
+import {rafDecorator} from '../../esl-utils/async/raf';
 import {ESLToggleable} from '../../esl-toggleable/core';
-import {ScrollUtils} from '../../esl-utils/dom/scroll';
 
-import type {ScrollStrategy} from '../../esl-utils/dom/scroll';
 import type {ToggleableActionParams} from '../../esl-toggleable/core';
 
+export interface PopupActionParams extends ToggleableActionParams {
+  /** popup position relative to trigger */
+  position?: string;
+  /** popup behavior if it does not fit in the window */
+  behavior?: string;
+  /** offset in pixels from trigger element */
+  offsetTrigger?: number;
+  /** offset in pixels from the edges of the window */
+  offsetWindow?: number;
+}
+
 @ExportNs('Popup')
 export class ESLPopup extends ESLToggleable {
   public static is = 'esl-popup';
 
-  @attr({defaultValue: null}) public disableScroll: ScrollStrategy;
+  public $arrow: HTMLElement | null;
+
+  protected _offsetTrigger: number;
+  protected _offsetWindow: number;
+  protected _deferredUpdatePosition = rafDecorator(() => this._updatePosition());
+
+  @attr({defaultValue: 'top'}) public position: string;
+  @attr({defaultValue: 'fit'}) public behavior: string;
+
+  /** Default params to merge into passed action params */
+  @jsonAttr<PopupActionParams>({defaultValue: {
+    offsetTrigger: 3,
+    offsetWindow: 15
+  }})
+  public defaultParams: PopupActionParams;
+
+  @prop() public closeOnEsc = true;
+  @prop() public closeOnOutsideAction = true;
+
+  public connectedCallback() {
+    this.$arrow = this.querySelector('.esl-popup-arrow');
+
+    super.connectedCallback();
+  }
+
+  protected bindEvents() {
+    super.bindEvents();
+    window.addEventListener('resize', this._deferredUpdatePosition);
+  }
+
+  protected unbindEvents() {
+    super.unbindEvents();
+    window.removeEventListener('resize', this._deferredUpdatePosition);
+  }
+
+  protected get _windowWidth() {
+    return document.documentElement.clientWidth || document.body.clientWidth;
+  }
+
+  protected get _windowX() {
+    return window.scrollX || window.pageXOffset;
+  }
+
+  protected get _windowY() {
+    return window.scrollY || window.pageYOffset;
+  }
```

Comment:
Placeholder: Make utility under esl-utils for getting that values.

Reply:
Will do it next wave